File: day1/solution.py
import math

def parse_input(file_name):
    parsed = []
    with open(file_name, "r") as f:
        for line in f.readlines():
            num = int(line[1:])
            if line[0] == "L":
                num *= -1
            parsed.append(num)
    return parsed

def part1(file_name):
    ans = 0
    dial = 50
    for num in parse_input(file_name):
        dial += num
        dial %= 100
        if dial == 0:
            ans += 1
    return ans

def part2(file_name):
    ans = 0
    dial = 50
    for num in parse_input(file_name):
        if num > 0:
            ans += (dial+num) // 100 - dial // 100
        else:
            ans += (dial-1) // 100 - (dial-1+num) // 100
        dial += num
        dial %= 100

    return ans

if __name__ == "__main__":
    print(part1("input.txt"))
    print(part2("input.txt"))

File: day1/solution.c
#include <stdio.h>
#include <stdlib.h>

int* parse_input(char *file_name, int *num_lines) {
    FILE *fptr = fopen(file_name, "r");

    char dir;
    int num;
    while (fscanf(fptr, "%c%d\n", &dir, &num) != EOF) {
        (*num_lines)++;
    }

    rewind(fptr);
    
    int *input = malloc(sizeof(int) * *num_lines);
    for (int i = 0; i < *num_lines; i++) {
        fscanf(fptr, "%c%d\n", &dir, &num);
        if (dir == 'L') {
            num *= -1;
        }
        input[i] = num;
    }
    fclose(fptr);
    return input;
}

int part1(char *file_name) {
    int num_lines = 0;
    int *input = parse_input(file_name, &num_lines);

    int ans = 0;
    int dial = 50;
    for (int i = 0; i < num_lines; i++) {
        dial += input[i];
        dial = (dial % 100 + 100) % 100;
        if (dial == 0) {
            ans += 1;
        }
    }
    return ans;
}

int part2(char *file_name) {
    int num_lines = 0;
    int *input = parse_input(file_name, &num_lines);

    int ans = 0;
    int dial = 50;
    for (int i = 0; i < num_lines; i++)
    {
        int num = input[i];
        int dir = 1;
        if (num < 0) {
            num *= -1;
            dir = -1;
        }
        for (int j = 0; j < num; j++) {
            dial += dir;
            if (dial == -1) {
                dial = 99;
            } else if (dial == 100) {
                dial = 0;
            }
            if (dial == 0) {
                ans += 1;
            }
        }
    }
    return ans;
}

int main() {

    printf("%d\n", part1("input.txt"));
    printf("%d\n", part2("input.txt"));

    return 0;
}